In [1]:
import pandas as pd
import matplotlib as plt
import numpy as np
from math import exp, factorial, log

Question 1:

In [2]:
#a) finding likelihood function
def poisson_pmf(lambda_, k):
    return np.exp(-lambda_) * (lambda_**k) / np.math.factorial(k)

#using observations instead of k to make it obvious for myself what it means
def likelihood_function(lambda_, observations):
    likelihood = 1.00
    for obs in observations:
      likelihood *= poisson_pmf(lambda_, obs)
    return likelihood

In [3]:
#testing my function to see if it works
observations = [2, 3, 4, 5, 3, 2]
likelihood_function(1,observations)

5.9769294383351645e-09

In [4]:
def log_likelihood_function(lambda_,observations):
  log_likelihood = 0.0
  for obs in observations:
    log_likelihood += log(poisson_pmf(lambda_,obs))
  return log_likelihood

#testing my function
log_likelihood_function(1,observations)

-18.93535887270599

In [1]:
#for all poisson distributions, the maximum likelihood estimator lambda will be the mean of the observations!
#derivation for part c is on the next page
def MLE(observations):
    lambda_mle = sum(observations) / len(observations)
    return lambda_mle

Question 2:

In [6]:
new_observations = [22, 23, 23, 12, 18, 25, 23, 14]
MLE(new_observations)

20.0

In [7]:
def neg_log_likelihood(lambda_,observations):
  return -log_likelihood_function(lambda_,observations)

In [8]:
from scipy.optimize import minimize

In [9]:
new_observations = [22, 23, 23, 12, 18, 25, 23, 14]
lambda_guess = np.array([15.0], dtype=np.float64)
result = minimize(neg_log_likelihood,lambda_guess, args=(new_observations))
print(result.x[0])
# lambda is approximately 20! Same as our MLE estimator function from Q1

19.999999685083786


Question 3:
part a) and b) are handwritten

In [11]:





#c
sample_observations = np.array([1.11, 3.61, 0.66, 2.22, 1.85, 1.36, 0.50, 0.33, 3.56, 0.08])

def neg_LLF(a,observations):
  return -np.sum(np.log(2/a-(2*observations/a**2)))

In [12]:
result = minimize(neg_LLF,4,args=sample_observations).x
print(result)

[4.5750106]


Question 4:

In [39]:
#a)
roll_1 = np.array([5, 3, 9, 3, 8, 4, 7])
roll_2 = np.array([5, 3, 9, 3, 8, 4, 7, 10])
freq_a = np.array([0,6,4,3,2,1,1,1,1,1,0]) #remember, 0 in index for convenience
freq_b = np.array([0,3,3,2,2,2,2,2,2,1,1])
freq_e = np.array([0,2,2,2,2,2,2,2,2,2,2])

cond_a = freq_a/np.sum(freq_a)
cond_b = freq_b/np.sum(freq_b)
cond_e = freq_e/np.sum(freq_e)

array([0. , 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1])

In [206]:
import pandas as pd
import seaborn as sns

data = roll_1

p_a,p_b,p_e = [1/3], [1/3],[1/3]
for d in data:
    p_a.append(cond_a[d]*p_a[-1] / (cond_a[d]*p_a[-1] + cond_b[d]*p_b[-1]+cond_e[d]*p_e[-1]))
    p_b.append(cond_b[d]*p_b[-1] / (cond_a[d]*p_a[-1] + cond_b[d]*p_b[-1]+cond_e[d]*p_e[-1]))
    p_e.append(1-p_a[-1]-p_b[-1])
    # p_b.append(cond_dist_b[d]*p_b[-1] / (cond_dist_a[d]*p_a[-2] + cond_dist_b[d]*p_b[-1])) # note the -2

df = pd.DataFrame({'P(A|data)': p_a, 'P(B|data)': p_b,'P(E|data)': p_e})
df.loc[7]
df

P(A|data)  P(B|data)  P(E|data)
0   0.333333   0.333333   0.333333
1   0.200000   0.434783   0.365217
2   0.272727   0.359595   0.367678
3   0.199409   0.277817   0.522774
4   0.271994   0.229870   0.498136
5   0.157404   0.284948   0.557648
6   0.157404   0.284948   0.557648
7   0.085425   0.321863   0.592712

There is a 78% likelihood that the dice is B, and 22% likelihood that the dice is A.

In [15]:
#b)
data_2 = roll_2

p_a,p_b,p_e = [1/3], [1/3],[1/3]

for d in data_2:
  p_a.append(cond_a[d]*p_a[-1] / (cond_a[d]*p_a[-1] + cond_b[d]*p_b[-1]+cond_e[d]*p_e[-1]))
  p_b.append(cond_b[d]*p_b[-1] / (cond_a[d]*p_a[-1] + cond_b[d]*p_b[-1]+cond_e[d]*p_e[-1]))
  p_e.append(1-p_a[-1]-p_b[-1])

df_2 = pd.DataFrame({'P(A|data)': p_a, 'P(B|data)': p_b,'P(E|data)': p_e})
df_2

P(A|data)  P(B|data)  P(E|data)
0   0.333333   0.333333   0.333333
1   0.200000   0.434783   0.365217
2   0.272727   0.359595   0.367678
3   0.199409   0.277817   0.522774
4   0.271994   0.229870   0.498136
5   0.157404   0.284948   0.557648
6   0.157404   0.284948   0.557648
7   0.085425   0.321863   0.592712
8   0.000000   0.213538   0.786462

After roll 8 (rolling the 10) there is 21.35% chance that it is dice B vs. 78.65% chance that it is dice E.

c) dice A is eliminated from the pool of possibilities when calculating likelihood after rolling a 10 because we know that P(A|10)=0

In [204]:
#P(B|E)
def simulation_e(simulations):
  results = np.array([])
  for i in np.arange(simulations):
    rolls = np.array([])
    p_b = 0.5
    p_e = 0.5
    while p_e <= 0.99:
      roll = np.random.choice(np.arange(11), p=cond_e)
      rolls = np.append(rolls, roll)
      latest_roll = int(rolls[-1])
      p_b = cond_b[latest_roll] * p_b / (cond_b[latest_roll] * p_b + cond_e[latest_roll] * p_e)
      p_e = 1 - p_b
    results = np.append(results,len(rolls))
  return results.mean()

def simulation_b(simulations):
  results = np.array([])
  for i in np.arange(simulations):
    rolls = np.array([])
    p_b = 0.5
    p_e = 0.5
    while p_b <= 0.99:
      roll = np.random.choice(np.arange(11), p=cond_b)
      rolls = np.append(rolls, roll)
      latest_roll = int(rolls[-1])
      p_b = cond_b[latest_roll] * p_b / (cond_b[latest_roll] * p_b + cond_e[latest_roll] * p_e)
      p_e = 1 - p_b
    results = np.append(results,len(rolls))
  return results.mean()

In [199]:
simulation_e(1000)

82.197

In [205]:
simulation_b(1000)

92.875

Using a simulation of 1000 rolls: it takes around 80 rolls for us to determine that the dice is E with 99% probability given that the real dice has the probability distribution of dice E, and it takes around 90 rolls for us to determine that the dice is B with 99% probability given that the real dice has the probability distribution of dice B. In order to test the true mean, we would need to create a confidence interval of sorts, but that is outside the bounds of this question.

Question 5: by hand